python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,380 @@
+
+        self.model.eval()
+        with torch.no_grad():
+            for idx, (g, _) in enumerate(dataset):
+                feat = func_extract_feat(g)
+
+                logits = self.model(g, feat)
+                ori_pred_dict[idx] = logits.argmax(-1).data
+
+        # train the mask generator
+        for epoch in range(self.epochs):
+            loss = 0.0
+            pred_list = []
+
+            tmp = float(
+                self.init_tmp
+                * np.power(self.final_tmp / self.init_tmp, epoch / self.epochs)
+            )
+
+            self.elayers.train()
+            optimizer.zero_grad()
+
+            for idx, (g, _) in enumerate(dataset):
+                feat = func_extract_feat(g)
+
+                prob, edge_mask = self.explain_graph(
+                    g, feat, tmp=tmp, training=True
+                )
+
+                self.edge_mask = edge_mask
+
+                loss_tmp = self.loss(prob.unsqueeze(dim=0), ori_pred_dict[idx])
+                loss_tmp.backward()
+
+                loss += loss_tmp.item()
+                pred_label = prob.argmax(-1).item()
+                pred_list.append(pred_label)
+
+            optimizer.step()
+            print(f"Epoch: {epoch} | Loss: {loss}")
+
+    def explain_graph(self, graph, feat, tmp=1.0, training=False):
+        r"""Learn and return an an edge mask that play a crucial role to
+        explain the prediction made by the GNN for a graph. Also, return
+        the prediction made with just the edge mask.
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            A homogeneous graph.
+        feat : Tensor
+            The input feature of shape :math:`(N, D)`. :math:`N` is the
+            number of nodes, and :math:`D` is the feature size.
+        tmp : float
+            The temperature parameter fed to the sample procedure
+        training : bool
+            Indicates whether the concrete_sample is called during
+            training or evaluation.
+
+        Returns
+        -------
+        Tensor, Tensor
+            The classification probability for graph with edge mask,
+            the probability mask for graph edges.
+
+        Examples
+        --------
+
+        >>> import torch as th
+        >>> import torch.nn as nn
+        >>> import dgl
+        >>> from dgl.data import GINDataset
+        >>> from dgl.dataloading import GraphDataLoader
+        >>> from dgl.nn import GraphConv, PGExplainer
+
+        >>> # Define the model
+        >>> class Model(th.nn.Module):
+        ...     def __init__(self, in_feats, out_feats):
+        ...         super().__init__()
+        ...         self.conv = GraphConv(in_feats, out_feats)
+        ...         self.fc = nn.Linear(out_feats, 1)
+        ...         th.nn.init.xavier_uniform_(self.fc.weight)
+        ...
+        ...     def forward(self, g, h, embed=False, edge_weight=None):
+        ...         h = self.conv(g, emb, edge_weight=edge_weight)
```

Comment:
NameError: name 'emb' is not defined

Make sure users can run the example without errors.

Reply:
Fixed.